src/buildstream/sourcemirror.py

```diff
@@ -0,0 +1,203 @@
+not relevant to the SourceMirror plugin object and need not be implemented.
+
+Sources expose the following abstract methods. Unless explicitly mentioned,
+these methods are mandatory to implement.
+
+* :func:`SourceMirror.translate_url() <buildstream.source.SourceMirror.translate_url>`
+ 
+  Produce an appropriate URL for the given URL and alias.
+
+
+Class Reference
+---------------
+"""
+
+from typing import Optional, Dict, List, Any, TYPE_CHECKING
+
+from .node import MappingNode, SequenceNode
+from .plugin import Plugin
+from ._exceptions import BstError, LoadError
+from .exceptions import ErrorDomain, LoadErrorReason
+
+if TYPE_CHECKING:
+
+    # pylint: disable=cyclic-import
+    from ._context import Context
+    from ._project import Project
+
+    # pylint: enable=cyclic-import
+
+
+class SourceMirrorError(BstError):
+    """This exception should be raised by :class:`.SourceMirror` implementations
+    to report errors to the user.
+
+    Args:
+       message: The breif error description to report to the user
+       detail: A possibly multiline, more detailed error message
+       reason: An optional machine readable reason string, used for test cases
+
+    *Since: 2.2*
+    """
+
+    def __init__(
+        self, message: str, *, detail: Optional[str] = None, reason: Optional[str] = None, temporary: bool = False
+    ):
+        super().__init__(message, detail=detail, domain=ErrorDomain.SOURCE, reason=reason)
+
+
+class SourceMirror(Plugin):
+    """SourceMirror()
+
+    Base SourceMirror class.
+
+    All SourceMirror plugins derive from this class, this interface defines how
+    the core will be interacting with SourceMirror plugins.
+
+    *Since: 2.2*
+    """
+
+    # The SourceMirror plugin type is only supported since BuildStream 2.2
+    BST_MIN_VERSION = "2.2"
+
+    def __init__(
+        self,
+        context: "Context",
+        project: "Project",
+        node: MappingNode,
+    ):
+        # Note: the MappingNode passed here is already expanded with
+        #       the project level base variables, so there is no need
+        #       to expand them redundantly here.
+        #
+        node.validate_keys(["name", "kind", "config", "aliases"])
+
+        # Do local base class parsing first
+        name: str = node.get_str("name")
+        self.__aliases: Dict[str, List[str]] = self.__load_aliases(node)
+
+        # Chain up to Plugin
+        super().__init__(name, context, project, node, "source-mirror")
+
+        # Plugin specific parsing
+        config = node.get_mapping("config", default={})
+        self._configure(config)
+
+    ##########################################################
+    #                        Public API                      #
+    ##########################################################
+    def translate_url(
+        self,
+        *,
+        project_name: str,
+        alias: str,
+        alias_url: str,
```

Comment:
Could it be sensible to make these two alias parameters optional? They are currently always set but we could consider supporting URL translation without an alias in the future. Making them optional in the signature now may make it easier to support this in the future without breaking backward compatibility.

Reply:
I actually spent some time inquiring on this.

Having optional keyword parameters does not work with normal python, unless the plugin author uses `**kwargs` syntax to declare the function, this won't be very pretty and have nice type annotations. 

If we instruct the plugin author to _only_ ever declare the function with `**kwargs` rather than explicitly typing out the parameters, then we can have functions with extensible keyword arguments in the API without breaking API.

Alternatively, we can have some internal monstrosity which observes the python function in memory only only feeds the function with parameters it is prepared to accept, which might look more pretty in documentation but wont have any sensible type checking with mypy either.

Keyword arguments are nice and extensible for us when we _provide_ a function that plugins _may call_, but not the other way around, we could go ahead with instructing plugin authors to use a `def function(self, **kwargs)` only, if we want extensibility, I'm not opposed to that but it does deviate from our current API style.

Reply:
With optional I meant marking the type as `Optional[str]` for `alias` and `alias_url` as we already do for `alias_substitute_url`. BuildStream core could in the future call it with `None` if we add support for non-alias translation. This would have to be opt-in for source mirror plugins.